### PR TITLE
Completed Basic Authorization with Logging functionality

### PR DESCRIPTION
**Pull request by ranizilpelwar**
## Pivotal Link
https://www.pivotaltracker.com/story/show/159180570

## Description 

**ResponseBodyBuilder**
- Added methods to include log content into response body when /logs is called with valid credentials
  
**BasicAuthorizationHandler**
- Added class to utilize Cipher, Salt and ability to validate the incoming credentials when logs are requested. Instead of storing the credentials in the code, I decided to store the encrypted credentials and to read from the config file called "config_settings.txt" in the config folder, if it exists, otherwise use the hash that contains the encrypted value. (Rather than storing the raw credentials or the actual credentials that gets you in). 

**Logger**  
- Added this class to log the first line of the Request Header into a file called "logs.txt".

**HTTPServerManager**  
- Added the ability to log the incoming requests if this is live code by including a new constructor that has a boolean. If it is true, it will log it, if it is not or not used, it won't log - such as when all the unit tests are being executed.
- I added try finally blocks wherever there were streams so that if something goes wrong, it will close the stream for sure, to avoid intermittent/erratic behavior between request processing.

### CODE REVIEW

File: src/main/java/com/server/RequestParser.java [hunk 1]
```diff
@@ -134,18 +135,33 @@ private int extractContentLength(){
...
     }
 
+    private String extractAuthorizationCredentials() {
+        String authorizationCredentials = "";
+        String authorizationHeaderLine = "Authorization: Basic";
+        if (headerString.contains(authorizationHeaderLine)) {
+            String[] headers = headerString.split("\r\n");
+            for ( String headerLine : headers) {
+                if (headerLine.contains(authorizationHeaderLine)) {
+                    authorizationCredentials = headerLine.replace(authorizationHeaderLine + " ", "");
```
Comment (shibani) [inline src/main/java/com/server/RequestParser.java]: Could we use the `extractLine` method here to extract the `authorizationHeaderLine` and then use the returned output to perform the `.replace`? similar to the `extractCookies` method?

File: src/main/java/com/server/ResponseBodyBuilder.java [hunk 2]
```diff
@@ -132,19 +133,22 @@ private String htmlBuilder() {
...
+            outputStream = new ByteArrayOutputStream();
+            int byteToBeRead = -1;
+            while ((byteToBeRead = inputStream.read()) != -1) {
+                outputStream.write(byteToBeRead);
+            }
+            imageContents = outputStream.toByteArray();
+            outputStream.flush();
+        }
+        finally {
+            outputStream.close();
```
Comment (shibani) [inline src/main/java/com/server/ResponseBodyBuilder.java]: looks great!

File: src/main/java/com/server/RequestParser.java (hunk 1 above)
Comment (ranizilpelwar) [reply src/main/java/com/server/RequestParser.java]: Refactored extractAuthorizationCredentials to use existing extractLine method

File: src/main/java/com/server/HTTPServerManager.java [hunk 3]
```diff
@@ -7,38 +7,47 @@
...
+            ServerSocket serverSocket = null;
+            Socket clientSocket = null;
+            try {
+                serverSocket = createServerSocket();
+                clientSocket = openSocket(serverSocket);
+
+                BufferedReader in = openInputStream(clientSocket);
+
+                RequestReader requestReader = new RequestReader(in);
+                String header = requestReader.getRequest();
```
Comment (Thomascountz) [inline src/main/java/com/server/HTTPServerManager.java]: Does `requestReader.getRequest()` return the entire request, i.e., all headers and body?

**Comment by ranizilpelwar** [comment]
Add method to return first line of header now that RequestParser is returning the full request and deleted some logs files that were generated during my troubleshooting, not the tests

**Pull request merged**